lib/Target/AVR/AVRInstrInfo.cpp

```diff
@@ -537,8 +537,7 @@ bool AVRInstrInfo::isBranchOffsetInRange(unsigned BranchOp,
     llvm_unreachable("unexpected opcode!");
   case AVR::JMPk:
   case AVR::CALLk:
-    assert(BrOffset >= 0 && "offset must be absolute address");
-    return isUIntN(16, BrOffset);
+    return true;
```

Comment:
I think this should work, since these instructions should always be able to reach any address, from any address.  
Regardless, the previous `assert` appears incorrect: `BrOffset` is negative for all jumps/calls where the destination address is lower to the address of the `JMP`/`CALL` instruction, which should certainly be valid.

Reply:
I don't think that is quite true

AVR supports addressing of >16-bit pointers through the use of extended IO instructions. In these cases, bits can be written to a special IO register which will then be taken to be the high bits of the address for an IO operation.

Also, LLVM will likely generate >16kb branches if the program is big enough. For example, if we had 16kb of NOPs and a call to a symbol located afterwards, LLVM might possibly make it all the way to this line before erroring.

Although I do have a feeling that we probably have an assert for this somewhere earlier in the pipeline, I think this assertion should stay here

Reply:
[Here's](https://imagecraft.com/help/iccavr/wwhelp/wwhimpl/common/html/wwhelp.htm?context=ICCAVRHelp&file=6-ProgrammingAVR12.html) some documentation on the extended IO feature

Reply:
I'll read up on that and test bigger branches, but note that I didn't remove those lines without reason. If the logic needs to remain, they need fixing, and I'm not sure how.

With the call to `isUIntN` left in, the code enters an infinite loop that uses up all RAM by creating hundreds of thousands of basic blocks, as `isBranchOffsetInRange` returns false when inserting a backwards jump (`BrOffset` negative), which makes it insert another.  
With the `assert` left in, we get an assertion failure any time we jump to a lower PC.

Reply:
Are you sure that's relevant? It looks to be about data (including LPM) access only, not instructions.  
JMP should, according to the AVR Instruction Set Manual, be able to reach any location on any AVR (up 4M words of memory), as the address is stored as 22 bits.

As for CALL, the manual shows it as having two forms, for 16-bit vs 22-bit PC, but there's only one opcode, with 22 bits.  
Each versions handles the entire program memory for the devices that has that instruction form,128 kB for devices with 128 kB and less, up to 8 MB for devices with more than 128 kB).

Reply:
It seems like I'm crossing my wires!

I believe you are correct, nice investigating